espnet2/asr_transducer/beam_search_transducer.py

```diff
@@ -0,0 +1,638 @@
+        self.beam_size = beam_size
+
+        if search_type == "default":
+            self.search_algorithm = self.default_beam_search
+        elif search_type == "tsd":
+            assert max_sym_exp > 1, "max_sym_exp should be greater than one."
+            self.max_sym_exp = max_sym_exp
+
+            self.search_algorithm = self.time_sync_decoding
+        elif search_type == "alsd":
+            assert u_max >= 0, "u_max should be a positive integer, a portion of max_T."
+            self.u_max = u_max
+
+            self.search_algorithm = self.align_length_sync_decoding
+        elif search_type == "maes":
+            assert self.dim_vocab >= beam_size + expansion_beta, (
+                "beam_size (%d) + expansion_beta (%d) "
+                " should be smaller or equal to vocab size (%d)."
+                % (beam_size, expansion_beta, self.dim_vocab)
+            )
+            self.max_candidates = beam_size + expansion_beta
+
+            self.nstep = nstep
+            self.expansion_gamma = expansion_gamma
+
+            self.search_algorithm = self.modified_adaptive_expansion_search
+        else:
+            raise NotImplementedError(
+                "Provided search type (%s) is not supported." % search_type
+            )
+
+        self.use_lm = lm is not None
+        self.lm = lm
+        self.lm_weight = lm_weight
+
+        self.score_norm = score_norm
+        self.nbest = nbest
+
+    def __call__(
+        self, enc_out: torch.Tensor
+    ) -> Union[List[Hypothesis], List[ExtendedHypothesis]]:
+        """Perform beam search.
+
+        Args:
+            enc_out: Encoder output sequence. (T, D_enc)
+
+        Returns:
+            nbest_hyps: N-best decoding results
+
+        """
+        self.decoder.set_device(enc_out.device)
+
+        nbest_hyps = self.search_algorithm(enc_out)
+
+        return nbest_hyps
+
+    def sort_nbest(
+        self, hyps: Union[List[Hypothesis], List[ExtendedHypothesis]]
+    ) -> Union[List[Hypothesis], List[ExtendedHypothesis]]:
+        """Sort hypotheses by score or score given sequence length.
+
+        Args:
+            hyps: Hypothesis.
+
+        Return:
+            hyps: Sorted hypothesis.
+
+        """
+        if self.score_norm:
+            hyps.sort(key=lambda x: x.score / len(x.yseq), reverse=True)
+        else:
+            hyps.sort(key=lambda x: x.score, reverse=True)
+
+        return hyps[: self.nbest]
+
+    def recombine_hyps(self, hyps: List[Hypothesis]) -> List[Hypothesis]:
+        """Recombine hypotheses with same label ID sequence.
+
+        Args:
+            hyps: Hypotheses.
+
+        Returns:
+            final: Recombined hypotheses.
+
+        """
+        final = []
+
+        for hyp in hyps:
+            seq_final = [f.yseq for f in final if f.yseq]
+
+            if hyp.yseq in seq_final:
+                seq_pos = seq_final.index(hyp.yseq)
+
+                final[seq_pos].score = np.logaddexp(final[seq_pos].score, hyp.score)
```

Comment:
Have you tried `max` instead of `log-sum-exp`?

Reply:
I did try max but it yields slightly worse performance in my experiments. What's the current conclusion on your side?

Reply:
See https://github.com/k2-fsa/icefall/pull/420
and
https://github.com/k2-fsa/icefall/pull/277

log-add also performs better than max in our experiments